File: backend/services/post-ts-koa-mongo-fdm/src/user-login/impl.ts
import { UserName, UserId } from "../common";
import { Deps } from "../common/utils";
import { Workflow } from ".";
import { WithId } from "mongodb";
import * as db from "../common/utils/db";
import { errors } from "./api";
import * as runtypes from "runtypes";
import { PasswordVerifier } from "../common/utils/password";

export class WorkflowImpl extends Workflow {
    private static readonly userHasPassword = runtypes.Record({
        encryptedPassword: runtypes.String,
    });
    async getUserIdAndPasswordVerifier(userName: UserName): Promise<[UserId, PasswordVerifier]> {
        const rec: WithId<unknown> | null = await this.deps.mongoDb.collection(db.users).findOne({ name: userName } satisfies { name: string });
        if (rec === null) {
            throw this.errors.userNameOrPasswordIncorrect();
        }
        db.validate(WorkflowImpl.userHasPassword, db.users, rec);
        return [db.formatId(rec._id) as UserId, new db.BCryptVerifier(rec.encryptedPassword)];
    }

    readonly errors = errors;
    constructor(private readonly deps: Deps) { super(); }
}

File: backend/services/post-kt-vertx-fdm/src/main/kotlin/me/jason5lee/post_kt_vertx_fdm/user_login/impl.kt
package me.jason5lee.post_kt_vertx_fdm.user_login

import com.mongodb.client.model.Filters
import com.mongodb.client.model.Projections
import me.jason5lee.post_kt_vertx_fdm.common.Password
import me.jason5lee.post_kt_vertx_fdm.common.UserId
import me.jason5lee.post_kt_vertx_fdm.common.UserName
import me.jason5lee.post_ktor_mongo_fdm.common.utils.Db
import me.jason5lee.post_ktor_mongo_fdm.common.utils.Deps
import org.bson.types.ObjectId

class WorkflowImpl(private val deps: Deps) : Workflow(), ErrorsImpl {
    override suspend fun verifyPasswordAndGetUserId(userName: me.jason5lee.post_kt_vertx_fdm.common.UserName, password: me.jason5lee.post_kt_vertx_fdm.common.Password): me.jason5lee.post_kt_vertx_fdm.common.UserId {
        val userDoc = Db.findOne<ObjectId>(
            db = deps.mongoDb,
            collection = Db.users,
            configCollection = {
                this.find(Filters.eq("name", userName.value))
                    .projection(Projections.include("encryptedPassword", "_id"))
            }
        ) ?: throw userNameOrPasswordIncorrect()
        val encryptedPassword = userDoc.get<String>("encryptedPassword")

        if (!password.verify(Db.BCrypt.Verifier(encryptedPassword))) {
            throw userNameOrPasswordIncorrect()
        }
        return me.jason5lee.post_kt_vertx_fdm.common.UserId(Db.formatId(userDoc.get("_id")))
    }
}
